Add support for initial velocity in MetaDrive simulations

lola831 opened the pull request:
### Description
This PR adds the ability to set an initial velocity for vehicles in the MetaDrive simulator.

### Issue Link
N/A

### Checklist
- [x] I have tested the changes locally via `pytest` and/or other means
- [ ] I have added or updated relevant documentation
- [x] I have autoformatted the code with black and isort
- [ ] I have added test cases (if applicable)

### Additional Notes
N/A

dfremont commented on src/scenic/simulators/metadrive/simulator.py:
```diff
@@ -112,6 +112,19 @@ def createObjectInSimulator(self, obj):
+                converted_heading,
+            ]
+            if obj.velocity and (obj.velocity.x != 0 or obj.velocity.y != 0):
+                full_md_velocity = utils.scenicToMetaDriveVelocity(obj.velocity)
```
This value isn't used: did you mean to use it in the next line?

lola831 replied:
Thanks for catching that—it's a leftover from my previous code and isn't needed.

dfremont commented on src/scenic/simulators/metadrive/simulator.py:
```diff
@@ -112,6 +112,19 @@ def createObjectInSimulator(self, obj):
+                converted_position,
+                converted_heading,
+            ]
+            if obj.velocity and (obj.velocity.x != 0 or obj.velocity.y != 0):
```
What's the purpose of this check? Is there a problem with setting the `spawn_velocity` to zero? Even if it's redundant to specify a zero velocity, I would do it just to simplify the code.

lola831 replied:
I've updated the code so that it always sets the initial velocity, even when it's zero. This change is included in commit 5907be30d64453b7a630880848253d1922ced824.

dfremont commented on src/scenic/simulators/metadrive/utils.py:
```diff
@@ -79,6 +79,26 @@ def metaDriveToScenicHeading(metaDriveHeading):
 
 
+def scenicToMetaDriveVelocity(scenic_velocity):
+    """Converts Scenic velocity to MetaDrive velocity by rotating the vector by π/2 (90 degrees)."""
```
This doesn't look right to me: MetaDrive's coordinate system is only shifted from Scenic's, not rotated (`scenicToMetaDrivePosition` only applies a constant offset). So you should not need to convert the velocity vector at all.

dfremont reviewed (changes requested):
A few questions and issues above. Also please add a test to make sure it works correctly: for example, you could create a car with a nonzero velocity but no behavior and check that it actually moves in the expected direction after one time step.

lola831 commented:
@dfremont Thanks for your comments. I've addressed the issues you raised and added a test to verify that a car with a nonzero initial velocity moves in the expected direction after one time step. These changes are included in commit 5907be30d64453b7a630880848253d1922ced824.

dfremont commented on tests/simulators/metadrive/test_metadrive.py:
```diff
@@ -128,3 +128,26 @@ def test_pedestrian_movement(getMetadriveSimulator):
+    initialPos = simulation.result.records["InitialPos"]
+    finalPos = simulation.result.records["FinalPos"]
+    dx = finalPos[0] - initialPos[0]
+    assert dx < 0, f"Expected car to move west (negative dx), but got dx = {dx}"
```
```suggestion
    assert dx < -0.1, f"Expected car to move west (negative dx), but got dx = {dx}"
```
Let's use a significant offset here so that small numerical errors can't cause this test to wrongly pass even if the velocity isn't being set properly.

dfremont reviewed (changes requested):
Looks great, just one tiny tweak above.

lola831 commented:
> Looks great, just one tiny tweak above.

Sounds good.
